### PR TITLE
add Laravel/framework support, both normal and facades

### PR DESCRIPTION
**Pull request by xiaohuilam**
image
image
image

normally, phpcd will return [facadesAccessor, __call...] method list, and it's not helpful completion.
this pull request is to solve this problem.

---
This change is [Reviewable](https://reviewable.io/reviews/lvht/phpcd.vim/105)

### CODE REVIEW

File: php/PHPCD.php [hunk 1]
```diff
@@ -1019,4 +1025,68 @@ public function classmap($pattern)
...
+
+        if (!is_readable($composer_path)) {
+            return [];
+        }
+
+        $composer = json_decode(file_get_contents($composer_path), true);
+        foreach(isset($composer['require'])?$composer['require']:[] as $package => $ver) {
+            if($package == 'laravel/framework') {
+                // 有使用 laravel 框架
+                // 增加 facades 支持
```
Comment (lvht) [inline php/PHPCD.php]: no chinese comments, please. thanks.

File: php/PHPCD.php [hunk 2]
```diff
@@ -25,11 +25,14 @@ class PHPCD implements RpcHandler
...
     private $root;
 
+    private $faced_class = [];
+
     public function __construct($root, Logger $logger, int $disable_modifier = 0, $match_type = self::MATCH_HEAD)
     {
         $this->logger = $logger;
         $this->root = $root;
         $this->disable_modifier = $disable_modifier;
+        $this->addSupportForLaravel();
```
Comment (lvht) [inline php/PHPCD.php]: please add config to enable this feature. and disable it by default.

File: php/PHPCD.php [hunk 3]
```diff
@@ -1019,4 +1025,68 @@ public function classmap($pattern)
...
+        $composer = json_decode(file_get_contents($composer_path), true);
+        foreach(isset($composer['require'])?$composer['require']:[] as $package => $ver) {
+            if($package == 'laravel/framework') {
+                // 有使用 laravel 框架
+                // 增加 facades 支持
+                if(file_exists($this->root.'/config/app.php')) {
+                    $config = require_once($this->root.'/config/app.php');
+                    $aliases = $config['aliases'];
+                    $require = [
+                        'App' => 'Illuminate\Foundation\Application',
```
Comment (lvht) [inline php/PHPCD.php]: maybe we need a more generic alias support. not only for laravel framework but also for any aliased.

Comment (xiaohuilam) [reply php/PHPCD.php]: Laravel's facades is complex,
For example, DB might aliased to Illuminate\Database\DatabaseManager::class, but also can aliased to Illuminate\Database\Connection::class (but laravel provides a hidden class named Illuminate\Database\Capsule\Manager. [patch](https://github.com/lvht/phpcd.vim/pull/105/commits/2ad962902f3f182e80c0f99b146d4b0607fc29bc) ).

So, it's so hard to make a universal formula to revert facaded classes, even we made, it's hard to adapt all frameworks.

**Comment by lvht** [comment]
@xiaohuilam This feature could be added without modifying phpcd's codebase.

Please see the [wiki](https://github.com/lvht/phpcd.vim/wiki/Laravel-Facade-Support), and I will close this PR. Thanks for your effort.

**Pull request closed without merging**